pubsub/google/cloud/pubsub_v1/publisher/batch.py

```diff
@@ -0,0 +1,231 @@
+# Copyright 2017, Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import collections
+import copy
+import queue
+import time
+
+import six
+
+from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.publisher import exceptions
+from google.cloud.pubsub_v1.publisher import future
+
+
+class Batch(object):
+    """A batch of messages.
+
+    The batch is the internal group of messages which are either awaiting
+    publication or currently in-flight.
+
+    A batch is automatically created by the PublisherClient when the first
+    message to be published is received; subsequent messages are added to
+    that batch until the process of actual publishing _starts_.
+
+    Once this occurs, any new messages sent to ``publish`` open a new batch.
+
+    If you are using this library, you most likely do not need to instantiate
+    batch objects directly; they will be created for you. If you want to
+    change the actual batching settings, see the ``batching`` argument on
+    :class:`google.cloud.pubsub_v1.PublisherClient`.
+
+    Args:
+        client (:class:`google.cloud.pubsub_v1.PublisherClient`): The
+            publisher client used to create this batch. Batch settings are
+            inferred from this.
+        topic (str): The topic. The format for this is
+            ``projects/{project}/topics/{topic}``.
+        settings (:class:`google.cloud.pubsub_v1.types.Batching`): The
+            settings for batch publishing. These should be considered
+            immutable once the batch has been opened.
+        autocommit (bool): Whether to autocommit the batch when the time
+            has elapsed. Defaults to True unless ``settings.max_latency`` is
+            inf.
+    """
+    def __init__(self, client, topic, settings, autocommit=True):
+        self._client = client
+
+        # Create a namespace that is owned by the client manager; this
```

Comment:
I used a `multiprocessing.Manager` here (the actual manager lives on the client) as the way of being able to exchange data between processes. I am unsure of how well this would hold up if a different concurrency system was in use.